mathics/builtin/colors.py

```diff
@@ -0,0 +1,373 @@
+
+    cie_l_100 = cie_l * 100.0  # MMA specific
+    y = (cie_l_100 + 16.) / 116.
+    y = _scale_lab_to_xyz(y)
+
+    x = -(9 * y * u) / ((u - 4) * v - u * v)
+    z = (9 * y - (15 * v * y) - (v * x)) / (3 * v)
+
+    cie_l_is_zero = cie_l < 0.078
+    x, y, z = map(lambda t: _luv_to_xyz_clip_zero(cie_l_is_zero, t), (x, y, z))
+    x, y, z = clip([x, y, z], 0, 1)
+
+    return (x, y, z) + rest
+
+
+def lch_to_lab(l, c, h, *rest):
+    h *= 2 * pi  # MMA specific
+    return (l, c * cos(h), c * sin(h)) + rest
+
+
+@conditional
+def _wrap_lch_h(h, pi2):
+    if h < 0.:
+        return h + pi2
+    else:
+        return h
+
+
+def lab_to_lch(l, a, b, *rest):
+    h = _wrap_lch_h(arctan2(b, a), 2. * pi)
+    h /= 2. * pi  # MMA specific
+    return (l, sqrt(a * a + b * b), h) + rest
+
+
+def lab_to_xyz(l, a, b, *rest):
+    # see http://www.easyrgb.com/index.php?X=MATH&H=08#text8
+    f_y = (l * 100. + 16.) / 116.
+    x, y, z = a / 5. + f_y, f_y, f_y - b / 2.
+    x, y, z = map(_scale_lab_to_xyz, (x, y, z))
+
+    x, y, z = [u * v for u, v in zip((x, y, z), _ref_white.xyz)]
+    x, y, z = clip([x, y, z], 0, 1)
+
+    return (x, y, z) + rest
+
+
+_flows = {  # see http://www.brucelindbloom.com/Math.html
+    'XYZ>LCH': ('XYZ', 'LAB', 'LCH'),
+    'LAB>LUV': ('LAB', 'XYZ', 'LUV'),
+    'LAB>RGB': ('LAB', 'XYZ', 'RGB'),
+    'LCH>XYZ': ('LCH', 'LAB', 'XYZ'),
+    'LCH>LUV': ('LCH', 'LAB', 'XYZ', 'LUV'),
+    'LCH>RGB': ('LCH', 'LAB', 'XYZ', 'RGB'),
+    'LUV>LAB': ('LUV', 'XYZ', 'LAB'),
+    'LUV>LCH': ('LUV', 'XYZ', 'LAB', 'LCH'),
+    'LUV>RGB': ('LUV', 'XYZ', 'RGB'),
+    'RGB>LAB': ('RGB', 'XYZ', 'LAB'),
+    'RGB>LCH': ('RGB', 'XYZ', 'LAB', 'LCH'),
+    'RGB>LUV': ('RGB', 'XYZ', 'LUV'),
+}
+
+
+_rgb_flows = set(['Grayscale', 'CMYK', 'HSB'])
+
+
+def _flow(src, dst):
+    if (src in _rgb_flows and dst != 'RGB') or (dst in _rgb_flows and src != 'RGB'):
+        return list(chain(_flow(src, 'RGB'), _flow('RGB', dst)))
+    else:
+        r = _flows.get('%s>%s' % (src, dst))
+        return list(r) if r else [src, dst]
+
+
+conversions = {
+    'Grayscale>RGB': grayscale_to_rgb,
+    'RGB>Grayscale': rgb_to_grayscale,
+    'CMYK>RGB': cmyk_to_rgb,
+    'RGB>CMYK': rgb_to_cmyk,
+    'RGB>HSB': rgb_to_hsb,
+    'HSB>RGB': hsb_to_rgb,
+    'XYZ>LAB': xyz_to_lab,
+    'XYZ>LUV': xyz_to_luv,
+    'XYZ>RGB': xyz_to_rgb,
+    'LAB>XYZ': lab_to_xyz,
+    'LAB>LCH': lab_to_lch,
+    'LCH>LAB': lch_to_lab,
+    'LUV>XYZ': luv_to_xyz,
+    'RGB>XYZ': rgb_to_xyz,
+}
+
+
+def convert(components, src, dst, preserve_alpha=True):
+    if not preserve_alpha:
+        components = stacked(omit_alpha, components)
+
+    if src == dst:
+        return components
+
+    flows = _flow(src, dst)
+    for s, d in (flows[i:i + 2] for i in range(len(flows) - 1)):
+        if s == d:
+            continue
+        func = conversions.get('%s>%s' % (s, d))
+        if not func:
+            return None
+        components = stacked(func, components)
```

Comment:
`LCH -> LAB -> XYZ -> RGB -> HSB` would be handled by `flow` using `RGB` as proxy (first if case) and then converting `LCH` to `RGB` and `RGB` to `HSB`. The code is a mess though, and I'll redo this with a simple and properly precomputed graph route table.

The underlying graph is beautiful, with `RGB` and `XYZ` being the central conversion hubs:

<img width="591" alt="color-graph" src="https://cloud.githubusercontent.com/assets/11859538/18158337/f7abeb0a-7022-11e6-8b90-7d8d8cc569df.png">